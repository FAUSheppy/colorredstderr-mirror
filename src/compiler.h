/*
 * Compiler specific macros.
 *
 * Copyright (C) 2013-2014  Simon Ruderich
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef COMPILER_H
#define COMPILER_H 1

#ifdef HAVE___ATTRIBUTE__
/* Prevent/force inlining. Used to improve performance. */
# define noinline      __attribute__((noinline))
# define always_inline __attribute__((always_inline))
/* Unused parameter. */
# define unused        __attribute__((unused))
/* Mark the function protected, which means it can't be overwritten by other
 * modules (libraries), e.g. with LD_PRELOAD); otherwise same as the default
 * visibility. This causes the compiler not use the PLT (and no relocations)
 * for local calls from inside this module; but the symbols are still
 * exported. This is faster and in this case prevents useless lookups as we
 * hook those functions and nobody else should modify them. Not strictly
 * necessary, but nice to have. */
# define visibility_protected __attribute__((visibility("protected")))
#else
# define noinline
# define always_inline
# define unused
# define visibility_protected
#endif

/* Branch prediction information for the compiler. */
#ifdef HAVE___BUILTIN_EXPECT
# define likely(x)   __builtin_expect(!!(x), 1)
# define unlikely(x) __builtin_expect(!!(x), 0)
#else
# define likely(x)   x
# define unlikely(x) x
#endif

#endif
