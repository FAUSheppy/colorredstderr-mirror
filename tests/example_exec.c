/*
 * Test execve() and exec*() functions.
 *
 * Copyright (C) 2013  Simon Ruderich
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

/* For execvpe(). */
#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "example.h"
#include "../src/compiler.h"


#define MAGIC "@RUN_"
#define MAGIC_LENGTH (strlen(MAGIC))

#define SETUP_RUN(number) \
        sprintf(argv0, "%s" MAGIC "%03d", argv[0], (number))


extern char **environ;


static int find_magic_run_number(char *argv0, int *number) {
    size_t length = strlen(argv0);
    if (length <= MAGIC_LENGTH + 3) {
        return 0;
    }
    /* Check if the string ends with "@RUN_YYY" where YYY should be a number
     * (not checked). */
    argv0 += length - MAGIC_LENGTH - 3;
    if (strncmp(argv0, MAGIC, MAGIC_LENGTH)) {
        return 0;
    }
    *number = atoi(argv0 + MAGIC_LENGTH);

    /* Strip off "@RUN_YYY". */
    argv0[0] = 0;
    return 1;
}

static int cmp(void const *a, void const *b) {
    return strcmp(*(char * const *)a, *(char * const *)b);
}
static void dump(char *argv[]) {
    size_t i;

    i = 0;
    while (*argv) {
        printf("argv[%zu] = |%s|\n", i++, *argv++);
    }

    /* Order of environment variables is not defined, sort them for consistent
     * test results. */
    i = 0;
    char **env = environ;
    while (*env++) {
        i++;
    }
    qsort(environ, i, sizeof(*env), cmp);

    i = 0;
    env = environ;
    while (*env) {
        /* Skip LD_PRELOAD which contains system dependent values. */
        if (strncmp(*env, "LD_PRELOAD=", 11)) {
            printf("environ[%zu] = |%s|\n", i++, *env);
        }
        env++;
    }
    printf("\n");
    /* When not writing to stdout (e.g. redirection while testing), glibc is
     * buffering heavily. Flush to display the output before the exec*(). */
    fflush(stdout);
}


int main(int argc unused, char **argv) {
    char argv0[strlen(argv[0]) + MAGIC_LENGTH + 3 + 1];

    char *old_ldpreload = getenv("LD_PRELOAD");
    size_t ldpreload_length = strlen("LD_PRELOAD=") + 1;
    if (old_ldpreload) {
        ldpreload_length += strlen(old_ldpreload);
    }
    char ldpreload[ldpreload_length];
    strcpy(ldpreload, "LD_PRELOAD=");
    if (old_ldpreload) {
        strcat(ldpreload, old_ldpreload);
    }

    int run;

    /* First call. */
    if (!find_magic_run_number(argv[0], &run)) {
        SETUP_RUN(1);

        char *args[] = { argv0, NULL };
        char *envp[] = { ldpreload, NULL };

        execve(argv[0], args, envp);
        return EXIT_FAILURE;
    }

    /* Following calls, each testing an exec*() feature. */

    dump(argv);
    SETUP_RUN(run + 1);

    int skip = run - 1;


    /* Check that we update/insert the environment correctly. */

    if (!skip--) {
        printf("\nCHECKING COLORING.\n\n");
        fflush(stdout);

        xdup2(2, 3);

        char *args[] = { argv0, NULL };
        char *envp[] = { ldpreload, NULL };

        execve(argv[0], args, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        xdup2(2, 4);

        execl(argv[0], argv0, NULL);
        return EXIT_FAILURE;
    } else if (!skip--) {
        xdup2(2, 5);

        execlp(argv[0], argv0, NULL);
        return EXIT_FAILURE;
    } else if (!skip--) {
        xdup2(2, 6);

        char *envp[] = { ldpreload, NULL };

        execle(argv[0], argv0, NULL, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        xdup2(2, 7);

        /* Test closing a few descriptors. */
        close(5);
        close(6);

        char *args[] = { argv0, NULL };

        execv(argv[0], args);
        return EXIT_FAILURE;
    } else if (!skip--) {
        xdup2(2, 8);

        /* And a few more. */
        close(7);
        close(4);

        char *args[] = { argv0, NULL };

        execvp(argv[0], args);
        return EXIT_FAILURE;

    /* Test handling of COLORED_STDERR_FDS. */

    } else if (!skip--) {
        /* And the rest. */
        close(3);
        close(8);

        xdup2(2, 5);

        char *args[] = { argv0, NULL };
        char *envp[] = { ldpreload, "COLORED_STDERR_FDS=5,", NULL };

        execve(argv[0], args, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        char *args[] = { argv0, NULL };
        char *envp[] = { ldpreload, NULL };

        xdup2(5, 6);
        close(5);

        execve(argv[0], args, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        close(6);

        char *args[] = { argv0, NULL };
        setenv("COLORED_STDERR_FDS", "2,", 1);

        execv(argv[0], args);
        return EXIT_FAILURE;
    }


    /* And make sure our hooks don't change the behaviour. */

    /* execve(2) */
    if (!skip--) {
        printf("\nCHECKING TRANSPARENCY.\n\n");
        fflush(stdout);

        char *args[] = { argv0, NULL };
        char *envp[] = { ldpreload, NULL };

        execve(argv[0], args, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        char *args[] = { argv0, NULL };
        char *envp[] = { "TEST=42", ldpreload, NULL };

        execve(argv[0], args, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        char *args[] = { argv0, "foo", "bar", NULL };
        char *envp[] = { "TEST=43", "FOO=", ldpreload, NULL };

        execve(argv[0], args, envp);
        return EXIT_FAILURE;

    /* execl(3) */
    } else if (!skip--) {
        setenv("TEST", "44", 1);

        execl(argv[0], argv0, NULL);
        return EXIT_FAILURE;
    } else if (!skip--) {
        setenv("TEST", "45", 1);

        execl(argv[0], argv0, "foo", "bar", NULL);
        return EXIT_FAILURE;

    /* execp(3), but not testing the p(ath) part */
    } else if (!skip--) {
        setenv("TEST", "46", 1);

        execlp(argv[0], argv0, NULL);
        return EXIT_FAILURE;
    } else if (!skip--) {
        setenv("TEST", "47", 1);

        execlp(argv[0], argv0, "foo", "bar", NULL);
        return EXIT_FAILURE;

    /* execle(3) */
    } else if (!skip--) {
        char *envp[] = { ldpreload, NULL };

        execle(argv[0], argv0, NULL, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        char *envp[] = { "TEST=48", ldpreload, NULL };

        execle(argv[0], argv0, NULL, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        char *envp[] = { "TEST=49", "FOO=", ldpreload, NULL };

        execle(argv[0], argv0, "foo", "bar", NULL, envp);
        return EXIT_FAILURE;

    /* execv(3) */
    } else if (!skip--) {
        setenv("TEST", "50", 1);

        char *args[] = { argv0, NULL };

        execv(argv[0], args);
        return EXIT_FAILURE;
    } else if (!skip--) {
        setenv("TEST", "51", 1);

        char *args[] = { argv0, "foo", "bar", NULL };

        execv(argv[0], args);
        return EXIT_FAILURE;

    /* execvp(3), but not testing the p(ath) part */
    } else if (!skip--) {
        setenv("TEST", "52", 1);

        char *args[] = { argv0, NULL };

        execvp(argv[0], args);
        return EXIT_FAILURE;
    } else if (!skip--) {
        setenv("TEST", "53", 1);

        char *args[] = { argv0, "foo", "bar", NULL };

        execvp(argv[0], args);
        return EXIT_FAILURE;

#ifdef HAVE_EXECVPE
    /* execvpe(3), but not testing the p(ath) part */
    } else if (!skip--) {
        char *args[] = { argv0, NULL };
        char *envp[] = { "TEST=54", ldpreload, NULL };

        execvpe(argv[0], args, envp);
        return EXIT_FAILURE;
    } else if (!skip--) {
        char *args[] = { argv0, "foo", "bar", NULL };
        char *envp[] = { "TEST=55", ldpreload, NULL };

        execvpe(argv[0], args, envp);
        return EXIT_FAILURE;
#else
    /* Fake output to let the test pass. */
    } else if (!skip--) {
        puts("argv[0] = |./example_exec|");
        puts("environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|");
        puts("environ[1] = |TEST=54|");
        puts("");
        puts("argv[0] = |./example_exec|");
        puts("argv[1] = |foo|");
        puts("argv[2] = |bar|");
        puts("environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|");
        puts("environ[1] = |TEST=55|");
        puts("");
#endif
    }

    printf("Done.\n");
    return EXIT_SUCCESS;
}
