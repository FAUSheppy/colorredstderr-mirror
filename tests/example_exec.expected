argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|


CHECKING COLORING.

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,3,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,3,4,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,3,4,5,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,3,4,5,6,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,3,4,7,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,3,8,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_FDS=5,|
environ[1] = |COLORED_STDERR_PRIVATE_FDS=2,5,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=6,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_FDS=2,|
environ[1] = |COLORED_STDERR_PRIVATE_FDS=|


CHECKING TRANSPARENCY.

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |TEST=42|

argv[0] = |./example_exec|
argv[1] = |foo|
argv[2] = |bar|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=43|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=44|

argv[0] = |./example_exec|
argv[1] = |foo|
argv[2] = |bar|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=45|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=46|

argv[0] = |./example_exec|
argv[1] = |foo|
argv[2] = |bar|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=47|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |TEST=48|

argv[0] = |./example_exec|
argv[1] = |foo|
argv[2] = |bar|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=49|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=50|

argv[0] = |./example_exec|
argv[1] = |foo|
argv[2] = |bar|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=51|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=52|

argv[0] = |./example_exec|
argv[1] = |foo|
argv[2] = |bar|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |FOO=|
environ[2] = |TEST=53|

argv[0] = |./example_exec|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |TEST=54|

argv[0] = |./example_exec|
argv[1] = |foo|
argv[2] = |bar|
environ[0] = |COLORED_STDERR_PRIVATE_FDS=2,|
environ[1] = |TEST=55|

Done.
EOF
